### PR TITLE
TXM-1154 ensure Akamai-Reputation header in HC by including all non-b…

### PR DESCRIPTION
**Pull request by iteratoruk**
…lacklisted headers

Further to recent discussion on #team-platops around extracting addition of headers to header carrier from code, this change set introduces forwarding of all headers with a configurable blacklist via Play config.

### CODE REVIEW

File: src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala [hunk 1]
```diff
@@ -114,10 +120,25 @@ object HeaderCarrier {
...
       headers.get(HeaderNames.googleAnalyticUserId),
-      getDeviceId(cookies, headers)
+      getDeviceId(cookies, headers),
+      headers.get(HeaderNames.akamaiReputation).map(AkamaiReputation),
+      remainingHeaders(headers)
     )
   }
 
+  val blacklistedHeaders: Seq[String] = {
+    // TODO sensible config key name and default values
```
Comment (beyond-code-github) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala]: TODO still left here!

File: src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala [hunk 2]
```diff
@@ -114,10 +120,25 @@ object HeaderCarrier {
       headers.get(HeaderNames.trueClientPort),
       headers.get(HeaderNames.googleAnalyticTokenId),
       headers.get(HeaderNames.googleAnalyticUserId),
-      getDeviceId(cookies, headers)
+      getDeviceId(cookies, headers),
+      headers.get(HeaderNames.akamaiReputation).map(AkamaiReputation),
```
Comment (beyond-code-github) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala]: Do we want to hard code the akamaiReputation header or will this now be covered by the blacklist changes?

File: src/main/scala/uk/gov/hmrc/play/http/HeaderNames.scala [hunk 3]
```diff
@@ -20,20 +20,46 @@ object HeaderNames {
...
-  val trueClientIp = "True-Client-IP"
-  val trueClientPort = "True-Client-Port"
-  val token = "token"
-  val surrogate = "Surrogate"
-  val otacAuthorization = "Otac-Authorization"
-  val googleAnalyticTokenId = "ga-token"
-  val googleAnalyticUserId  = "ga-user-cookie-id"
-  val deviceID  = "deviceID" // not a typo, should be ID
+  /*
+   * this isn't ideal, but downstream apps still want to refer to typed header values
```
Comment (beyond-code-github) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderNames.scala]: I wonder if there's a better way of doing this...  maybe case objects or enumerations (sorry my scala isn't all that great)

Also is there a reason this is a map rather than a seq?

File: src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala [hunk 4]
```diff
@@ -33,12 +34,14 @@ case class HeaderCarrier(authorization: Option[Authorization] = None,
                          requestId: Option[RequestId] = None,
                          requestChain: RequestChain = RequestChain.init,
                          nsStamp: Long = System.nanoTime(),
-                         extraHeaders: Seq[(String, String)] = Seq(), 
+                         extraHeaders: Seq[(String, String)] = Seq(),
```
Comment (beyond-code-github) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala]: Hmmmm, this seems suspiciously similar to 'otherHeaders'... I wonder what it's used for

File: src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala (hunk 2 above)
Comment (iteratoruk) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala]: I think we do still want to force Akamai-Reputation because we have downstream functionality that depends on it. By sneaking it in to the hard-coded headers with this change, we ensure that it could never be "accidentally" blacklisted by potential future configuration, which would break the downstream functionality.

File: src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala (hunk 4 above)
Comment (iteratoruk) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala]: "extraHeaders" have been explicitly added by the calling client (i.e. http-verbs user). "otherHeaders" are the "all the rest that are not blacklisted" which are now being passed-through.

File: src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala (hunk 1 above)
Comment (iteratoruk) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala]: Any suggestions for sensible config key names gratefully received! Since this config is currently hypothetical and would be a platops concern, I was a little coy about naming it ...

File: src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala (hunk 2 above)
Comment (beyond-code-github) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala]: Welcome back :) Sounds good.

File: src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala (hunk 4 above)
Comment (beyond-code-github) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala]: Hmmm tricky. I'd like to clarify the name 'extraHeaders' but I'm wary of introducing a breaking change. The only other thing I can think of is to rename 'otherHeaders' to something like 'headersNotPresentInBlacklist' which is a bit wordy

File: src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala (hunk 1 above)
Comment (beyond-code-github) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala]: I'd probably call it 'httpHeadersBlacklist' (present tense rather than past) but to be honest your name is fine really. Can just remove the TODO if you're happy with the defaults.

File: src/main/scala/uk/gov/hmrc/play/http/HeaderNames.scala (hunk 3 above)
Comment (iteratoruk) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderNames.scala]: It's a map because HeaderNames, prior to the change, was a collection of key->value pairs, i.e. not tuples because the keys are unique. We could use a sequence with index based access but, personally, I thought that was a little less clear (and also broken by "accidental" re-ordering of the collection). I think case objects etc would be overkill for something this simple where all we wanted to do was to be able to retrieve a key set whilst maintaining the val references for backward compatibility with existing client code which might need them.

**Comment by beyond-code-github** [comment]
Merged :+1:
